Сделал refactoring

denistishuk opened the pull request:
Сделал refactoring

denistishuk committed:
Сделал refactoring

Сделал refactoring

denistishuk committed:
Метод FindMax

Идея метода была отсортировать список сделать реверс и вернуть первый элемент.
Как отсортировать числа вопросов нет.  Метод должен быть универсальным. как отсортировать строки, например. А если там будут просто объекты хранится... ?
Короче нужна пояснительная бригада

denistishuk committed:
Стек MyStack

Хотел реализовать и переписать методы как в документации. Но почему то в массив нули добавляются.

denistishuk committed:
Стек на основе массива

Вроде реализовал. Но с обработкой ошибок пока гавнокод.

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStack.java:
```diff
@@ -0,0 +1,145 @@
+
+import java.util.*;
+
+public class MyStack<E> /*extends AbstractList<E> */{
```
Лучше бесполезные закомментированные куски кода не оставлять

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStack.java:
```diff
@@ -0,0 +1,145 @@
+
+public class MyStack<E> /*extends AbstractList<E> */{
+      Object[] elementData;
+    protected transient int modCount = 0;
```
Эммм... Transient это что? Что-то для сериализации, типо чтобы поле не сериализировать?

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStack.java:
```diff
@@ -0,0 +1,145 @@
+import java.util.*;
+
+public class MyStack<E> /*extends AbstractList<E> */{
+      Object[] elementData;
```
В общем проанализируй модификаторы доступа, у тебя непонятно почему они плавают - elementData зачем-то package private, остальные поля protected. Главное правило - чем жестче, тем лучше. Если нет причин открывать поле, то делай private

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStack.java:
```diff
@@ -0,0 +1,145 @@
+//    }
+
+    public MyStack(int initialCapacity, int capacityIncrement) {
+        super();
```
Конструктор по умолчанию от класса Object будет вызван в любом случае, зачем явно вызывать?

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStack.java:
```diff
@@ -0,0 +1,145 @@
+
+    public MyStack(int initialCapacity, int capacityIncrement) {
+        super();
+        if (initialCapacity < 0)
```
Вот тут хорошее решение, ты просто не сможешь создать объект с неправильными данными. Это хороший дизайн

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStack.java:
```diff
@@ -0,0 +1,145 @@
+        this.capacityIncrement = capacityIncrement;
+    }
+
+    public MyStack(int initialCapacity) {
```
Тоже хорошее решение, использовать полиморфные конструкторы. Таким стеком пользоваться удобнее, чем когда все параметры обязательно нужно забивать

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStack.java:
```diff
@@ -0,0 +1,145 @@
+    public MyStack(Collection<? extends E> c) {
+        Object[] a = c.toArray();
+        elementCount = a.length;
+        if (c.getClass() == ArrayList.class) {
```
Тут непонятно. Метод toArray любую коллекцию преобразует в массив,зачем эта проверка и потом копирование массива?
Точнее, почему не копировать массив каждый раз вне зависимости от того ArrayList это или нет?

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStack.java:
```diff
@@ -0,0 +1,145 @@
+        }
+    }
+//================================================================
+    public E push(E item) {
```
Обычно метод push не возвращает элемент

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStack.java:
```diff
@@ -0,0 +1,145 @@
+        addElement(item);
+        return item;
+    }
+    public synchronized void addElement(E obj) {
```
Тут непонятно, почему просто не сделать один метод push?

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStack.java:
```diff
@@ -0,0 +1,145 @@
+        elementCount = s + 1;
+    }
+
+/*    private Object[] grow() {
```
Закомментированные куски кода, не круто бро

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStack.java:
```diff
@@ -0,0 +1,145 @@
+
+
+    public synchronized E pop() {
+        E       obj;
```
Что с форматированием? Ctrl+Alt+L

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStack.java:
```diff
@@ -0,0 +1,145 @@
+
+    public synchronized E pop() {
+        E       obj;
+        int     len = size();
```
у тебя же доступно поле elementCount в этом контексте, зачем использовать метод?

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStack.java:
```diff
@@ -0,0 +1,145 @@
+        int     len = size();
+
+        if (len == 0)
+            throw new EmptyStackException();
```
Здесь я бы все-таки возвращал null, либо сделал EmptyStackException checked. Так из сигнатуры метода был бы понятен контракт метода

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStack.java:
```diff
@@ -0,0 +1,145 @@
+        return elementAt(len - 1);
+    }
+
+    public synchronized void removeElementAt(int index) {
```
Вот это не должно быть public, зачем предоставлять такой апи в стэке? У стэка должно быть 3 публичных метода - pop, push и size. Как он внутри устроен, это уже детали, которые надо инкапсулировать

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStack.java:
```diff
@@ -0,0 +1,145 @@
+        }
+        int j = elementCount - index - 1;
+        if (j > 0) {
+            System.arraycopy(elementData, index + 1, elementData, index, j);
```
Тут неочевидно, копирование массива сам в себя, точно нормально работает? Просто в цикле пройтись и сдвинуть элементы наверх было бы более правильно

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStack.java:
```diff
@@ -0,0 +1,145 @@
+        elementData[elementCount] = null; /* to let gc do its work */
+    }
+
+    public synchronized E elementAt(int index) {
```
То же самое, этот апи не стоит предоставлять публично

denistishuk replied:
согласен. Но этот метод почему то публичный в библиотеке стека

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStackArr.java:
```diff
@@ -0,0 +1,130 @@
+    protected transient int modCount = 0;
+
+    // Конструктор для инициализации stack
+    MyStackArr(int initialCapacity)
```
package private, это точно то что нужно?

denistishuk replied:
понял. сначала не понял что такое package private. Потом перечитал твои комменты понял)
Понял, что сразу сделал сам а скопипастил, а мозгов не хватило самому. То что в классе MyStackArr дизайн гавно полное.

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStackArr.java:
```diff
@@ -0,0 +1,130 @@
+    private final Object[] elementData;
+    private int top;
+    private  int capacity;
+    protected transient int modCount = 0;
```
Для чего protected transient?

denistishuk replied:
transient - это если мы не хотим сериализовывать это поле
убрал и изменил на private.

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStackArr.java:
```diff
@@ -0,0 +1,130 @@
+        }
+    }
+    // Вспомогательная функция для добавления элемента `x` в stack
+    public E push (E item) // public void E push (E item) - ругается
```
Конечно ругается, потому что это не валидный синтаксис =) Либо void, либо E. void указывает, что метод не возвращает результат, а ты потом еще указываешь,что он возвращает E

texaskst replied:
Еще момент - ты меняешь состояние массива в этом методе, но не делаешь синхронизацию

denistishuk replied:
понял

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStackArr.java:
```diff
@@ -0,0 +1,130 @@
+            elementData = Arrays.copyOf(a, top, Object[].class);
+        }
+    }
+    // Вспомогательная функция для добавления элемента `x` в stack
```
Как это вспомогательная функция, это основная функция публичного апи этого класса. Без нее он не имеет смысла)

denistishuk replied:
согласен. так было в копипасте.

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStackArr.java:
```diff
@@ -0,0 +1,130 @@
+        if (isFull())
+        {
+            System.out.println("Overflow\nProgram Terminated\n");
+            System.exit(-1);
```
Вот это пиздец конечно)))))) Прикинь у тебя микросервис работает такой, тут в каком-то месте кто-то заполнил стэк и ты вырубаешь весь процесс

denistishuk replied:
убрал. создал свое исключение

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStackArr.java:
```diff
@@ -0,0 +1,130 @@
+        return item;
+    }
+
+    // Вспомогательная функция для извлечения верхнего элемента из stack
```
Почему вспомогательная то? Это основа апи

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStackArr.java:
```diff
@@ -0,0 +1,130 @@
+    // Вспомогательная функция для извлечения верхнего элемента из stack
+    public synchronized E pop()
+    {
+        //E       obj;
```
Коменты, коменты...

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStackArr.java:
```diff
@@ -0,0 +1,130 @@
+        if (isEmpty())
+        {
+            System.out.println("Underflow\nProgram Terminated");
+            System.exit(-1);
```
Полный пиздец, за такое тебя бы как минимум с позором выгнали

denistishuk replied:
согласен. Это я скопировал реализацию стека. и на этом примере попытался сделать класс с джинериками по заданию. Касательно исключений мне надо лучше разобраться какой тип применять. Ниже ты написал что там пиздец. Пытаюсь разобраться в чем.

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/MyStackArr.java:
```diff
@@ -0,0 +1,130 @@
+            return (E) elementData[top];
+        }
+        else {
+            return (E) new RuntimeException("Stack is empty");
```
Еще один полный пиздец. Попробуй воспроизвести этот кейс и понять что не так

denistishuk replied:
Тут я попытался метод exit (-1) рефактринг сделать и дал такое исключение. Я так понял что надо свое исключение создать как ты объяснял на лекции рефакторинга

texaskst replied:
Ты пытаешься привести RuntimeException к E. E это тип элемента стека, может быть String, Integer, или еще какой-нибудь объект. То есть ты делаешь
```
(String) new RuntimeException()
```
Это называется "натянуть сову на глобус". В этом вообще никакого смысла нет

denistishuk committed:
MyStack (2 ревизия)

Вроде работает.  С коллекциями еще не тестил

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/EmptyStackException.java:
```diff
@@ -0,0 +1,12 @@
+        return "Error. Stack is empty.";
+    }
+
+//    public EmptyStackException(String message) {
```
Зачем опять закомментированый код?

texaskst commented on src/main/java/kz/lakida/javacourse/tishuk/genericDemo/FullStackException.java:
```diff
@@ -0,0 +1,12 @@
+        return "Error. Stack is full.";
+    }
+
+//    public EmptyStackException(String message) {
```
То же самое

denistishuk committed:
Реализовал метод FindMax

Не получилось сделать именно через Comparable как в задании.
Получилось довести до рабочего состояния.
Сделай ролик с пояснением как что надо было делать. Как правильно метод использовать.
ПОтому что я через Object вызов метода сделал. ХЗ

anton-at-baraka committed:
Implement findMax with comparable generic type
